src/Workspaces/MSBuild/Core/MSBuild/BuildHostProcessManager.cs

```diff
@@ -63,58 +65,98 @@ public async Task<RemoteBuildHost> GetBuildHostWithFallbackAsync(string projectF
-
-                // We've subscribed to Disconnected, but if the process crashed before that point we might have not seen it
-                if (process.HasExited)
+                bool reload;
+                do
                 {
-                    buildHostProcess.LogProcessFailure();
-                    throw new Exception($"BuildHost process exited immediately with {process.ExitCode}");
-                }
+                    reload = false;
+
+                    var pipeName = Guid.NewGuid().ToString();
+                    var processStartInfo = CreateBuildHostStartInfo(buildHostKind, pipeName, dotnetPath);
+
+                    var process = Process.Start(processStartInfo);
+                    Contract.ThrowIfNull(process, "Process.Start failed to launch a process.");
+
+                    buildHostProcess = new BuildHostProcess(process, pipeName, _loggerFactory);
+                    buildHostProcess.Disconnected += BuildHostProcess_Disconnected;
+
+                    // We've subscribed to Disconnected, but if the process crashed before that point we might have not seen it
+                    if (process.HasExited)
+                    {
+                        buildHostProcess.LogProcessFailure();
+                        throw new Exception($"BuildHost process exited immediately with {process.ExitCode}");
+                    }
+
+                    // When running on .NET Core, we need to find the right SDK location that can load our project and restart the BuildHost if required.
+                    // When dotnetPath is null, the BuildHost is started with the default dotnet executable, which may not be the right one for the project.
+                    if (buildHostKind == BuildHostProcessKind.NetCore
+                        && projectOrSolutionFilePath is not null
+                        && dotnetPath is null)
+                    {
+                        // The BuildHost will be able to search through all the SDK install locations for a usable MSBuild instance.
+                        var msbuildLocation = await buildHostProcess.BuildHost.FindUsableMSBuildAsync(projectOrSolutionFilePath, cancellationToken).ConfigureAwait(false);
+                        if (msbuildLocation is not null && GetProcessPath() is { } processPath)
+                        {
+                            // The layout of the SDK is such that the dotnet executable is always at the same relative path from the MSBuild location.
+                            dotnetPath = Path.GetFullPath(Path.Combine(msbuildLocation.Path, $"../../{DotnetExecutable}"));
+                            if (dotnetPath is not null && processPath != dotnetPath)
+                            {
+                                // We need to relaunch the .NET BuildHost from a different dotnet instance.
+                                reload = true;
+                                await buildHostProcess.DisposeAsync().ConfigureAwait(false);
```

Comment:
It seems this logic needs to be outside the if (!_processes.TryGetValue(buildHostKind, out var buildHostProcess)) check; because if you were to load several projects at once, once that had a usable MSBuild for the project, we'll cache that in _processes, and later ones wouldn't check again. That said, I think the only way you'd hit this in practice is if you have a solution that spans more than one global.json which isn't really supported by anything.

Reply:
I wonder if we're better off doing something like:

```
public async Task<RemoteBuildHost> GetBuildHostAsync(BuildHostProcessKind buildHostKind, CancellationToken cancellationToken)
```

where BuildHostProcessKind becomes a poor-mans discriminated union like:

```
record BuildHostProcessKind();
record DotNetHostProcess(string? dotnetPath) : BuildHostProcessKind();
record NetFrameworkHostProcess() : BuildHostProcessKind();
```

and then that way it's clear that way the path is integral to the kind.

Reply:
My goal for this PR was just to better handle the single SDK BuildHost scenario.

I like your suggestion for supporting multiple .NET Core build hosts. However, I think adding support that should be in a follow up.